app/main/log-utils.js

```diff
@@ -0,0 +1,86 @@
+/*
+ Copyright 2016 Mozilla
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+import path from 'path';
+import fs from 'fs-promise';
+
+/**
+ * This file is loaded in app/shared/logging via `tofino-log-utils` alias,
+ * since we don't want to include node APIs in the renderer process.
+ */
+
+const PATHS = {
+  linux: env => [
+    [env['XDG_CONFIG_HOME']],
+    [env['HOME'], '.config'],
+    [env['XDG_DATA_HOME']],
+    [env['HOME'], '.local', '.share'],
+  ],
+
+  darwin: env => [
+    [env['HOME'], 'Library', 'Logs'],
+    [env['HOME'], 'Library', 'Application Support'],
+  ],
+
+  win32: env => [
+    [env['APPDATA']],
+    [env['USERPROFILE'], 'AppData', 'Roaming'],
+  ],
+};
+
+function getFirstAvailableDirectory(dirs, appName) {
+  for (const dirPathParts of dirs) {
+    // If we have a part that's undefined, like from an undefined env var,
+    // this path cannot be valid
+    if (!dirPathParts.every(part => part !== undefined)) {
+      continue;
+    }
+
+    const dirPath = path.join(...dirPathParts, appName);
+    fs.mkdirpSync(dirPath);
+    try {
+      fs.accessSync(dirPath, fs.W_OK);
+    } catch (e) {
+      continue;
+    }
+    if (!fs.statSync(dirPath).isDirectory()) {
+      continue;
+    }
+
+    return dirPath;
+  }
+
+  return null;
+}
+
+/**
+ * Logic and default paths for platforms inspired by electron-log
+ * https://github.com/megahertz/electron-log
+ */
+export const getLogDirectory = function({
```

Comment:
This was originally async for fs access, but then the logger would have to buffer input as it's included in all environments synchronously. Should be async in the future so we're not doing this at startup, fine for now IMO